From: Paolo Bonzini <pbonzini@redhat.com>
Date: Wed, 23 Jun 2010 14:07:57 -0400
Subject: [virt] xen: remove sysdata hack from irq-xen.c
Message-id: <1277302078-10425-3-git-send-email-pbonzini@redhat.com>
Patchwork-id: 26468
O-Subject: [RHEL5.5 PATCH 2/3] xen: remove sysdata hack from
	arch/i386/pci/irq-xen.c
Bugzilla: 561390
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

Bugzilla: 561390

Brew build: https://brewweb.devel.redhat.com/taskinfo?taskID=2540880

This forward-ports the patch for BZ519633 so that it is available
in xen-enabled kernels too.

The code between irq.c and irq-xen.c differs already because of the
BZ470202 fix, which is actually the same bug as BZ519633 and whose patch
touched arch/i386/pci/irq-xen.c and arch/i386/pci/legacy.c.  The patch
for BZ519633 then introduced a function to do the same thing as this fix,
but only used it in legacy.c, not in irq-xen.c.

This patch finishes the conversion.  Reverting BZ470202 + applying
BZ519633 to match irq.c gives exactly the same result.

diff --git a/arch/i386/pci/irq-xen.c b/arch/i386/pci/irq-xen.c
index 9771540..a8c29c0 100644
--- a/arch/i386/pci/irq-xen.c
+++ b/arch/i386/pci/irq-xen.c
@@ -141,20 +141,11 @@ static void __init pirq_peer_trick(void)
 		busmap[e->bus] = 1;
 	}
 	for(i = 1; i < 256; i++) {
-		struct pci_sysdata *sd;
 		if (!busmap[i] || pci_find_bus(0, i))
 			continue;
-		/* Continuation of the gross hack.
-		   sysdata cannot be NULL here because of PCI_DOMAIN support.
-		   Let's assume we're part of the same domain and node */
-		sd = kzalloc(sizeof(*sd), GFP_KERNEL);
-		if (!sd)
-			panic("Cannot allocate PCI domain sysdata");
-		if (pci_scan_bus(i, &pci_root_ops, sd))
-			printk(KERN_INFO "PCI: Discovered primary peer bus %02x [IRQ]\n", i);
-		else
-			kfree(sd);
-
+		if (pci_scan_bus_with_sysdata(i))
+			printk(KERN_INFO "PCI: Discovered primary peer "
+			       "bus %02x [IRQ]\n", i);
 	}
 	pcibios_last_bus = -1;
 }
